lib/ansible/module_utils/api.py

```diff
@@ -114,3 +115,53 @@ def retried(*args, **kwargs):
 
         return retried
     return wrapper
+
+
+def generate_jittered_backoff(retries=10, delay=3, max_delay=60):
+    """The "Full Jitter" backoff strategy described here https://www.awsarchitectureblog.com/2015/03/backoff.html
+
+    :param retries: The number of delays to generate.
+    :param delay: The time in seconds for the initial delay.
+    :param max_delay: The maximum time in seconds for any delay.
+    """
+    for retry in range(0, retries):
+        yield random.randint(0, min(max_delay, delay * 2 ** retry))
+
+
+def retry_with_delays_and_condition(backoff_iterator, retry_condition, do_on_retry=None, retry_on_success=None):
+    """Generic retry decorator.
+
+    :param backoff_iterator: An iterable of delays in seconds.
+    :param retry_condition: A callable that takes an exception of the decorated function and decides whether to retry or not (returns a bool).
+    :param do_on_retry: An optional callable that takes an exception/None and the current delay and runs before the sleep (returns None).
+    :param retry_on_success: An optional callable that accepts the decorated function return value and decides whether or not to retry (returns a bool).
```

Comment:
I think I'd call it `retry_or_not` or `should_retry`. This way it wouldn't only refer to a positive boolean outcome.

Reply:
Do you think `retry_or_not`/`should_retry` would sound like it should run regardless of whether the function call was successful though? This callable should run if the decorated function doesn't raise an exception.

Reply:
That's an interesting observation. Maybe use `should_retry_on_error` then? Thinking about it, this is probably why backoff called their args `giveup` and `predicate` — the first one clearly relates to an error caught (and an expectation that something failed and there was no successful result which is a typical pythonic idiom) and the latter pretty obviously inspects the return value in a scenario when the errors aren't communicated via exceptions.

Reply:
I've renamed both of the retry conditions to `should_retry_error` and `should_retry_result`, lmkwyt